### PR TITLE
cmake: BuildFIO.cmake should not introduce -std=gnu++17

### PR DESCRIPTION
**Pull request by mattbenjamin**
Not correct in general, and a build bug because fio-objectstore includes c++20 headers.

Fixes: https://tracker.ceph.com/issues/62778

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: cmake/modules/BuildFIO.cmake [hunk 1]
```diff
@@ -38,5 +38,5 @@ function(build_fio)
   add_dependencies(fio fio_ext)
   set_target_properties(fio PROPERTIES
     INTERFACE_INCLUDE_DIRECTORIES ${source_dir}
-    INTERFACE_COMPILE_OPTIONS "-include;${source_dir}/config-host.h;$<$<COMPILE_LANGUAGE:C>:-std=gnu99>$<$<COMPILE_LANGUAGE:CXX>:-std=gnu++17>")
+    INTERFACE_COMPILE_OPTIONS "-include;${source_dir}/config-host.h;$<$<COMPILE_LANGUAGE:C>:-std=gnu99>$<$<COMPILE_LANGUAGE:CXX>:>")
```
Comment (tchaikov) [inline cmake/modules/BuildFIO.cmake:41]: the commit where i introduced these interface options was c190fee53d23212faf22b26df33346b393e11b8d. i wanted to populate `CXX_EXTENSIONS=ON` to the targets linking against `fio`. could you please preserve this ? probably we can use 

```cmake
set_target_properties(fio PROPERTIES
  CXX_EXTENSIONS ON)
```
?
`$<$<COMPILE_LANGUAGE:CXX>:>` is a generator expression, which expanded to an empty string. if we want top drop `-std=gnu++17`, we might want to drop the whole expression.

Comment (mattbenjamin) [reply cmake/modules/BuildFIO.cmake:41]: ok, will update;  thanks Kefu!

**Comment by tchaikov** [comment]
lgmt. could you squash these two commits?

**Comment by mattbenjamin** [comment]
sure

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/crushtool/set-choose.t: failed
--- /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/crushtool/set-choose.t
+++ /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/crushtool/set-choose.t.err
@@ -21261,3365 +21261,6 @@
   CRUSH rule 4 x 743 [7,2]
   CRUSH rule 4 x 744 [5,7]
   CRUSH rule 4 x 745 [5]
-  CRUSH rule 4 x 746 [5,2]
-  CRUSH rule 4 x 747 [6,2]
-  CRUSH rule 4 x 748 [2,7]
-  CRUSH rule 4 x 749 [5]
-  CRUSH rule 4 x 750 [2,6]
...
```

searched on tracker, turns out this is a new one. i don't think it is related. but better off understanding what's going on..

**Comment by tchaikov** [comment]
@mattbenjamin hi Matt, could you push this change to ceph-ci, so we can compile-test on more distros?

**Comment by idryomov** [comment]
jenkins test make check

**Comment by mattbenjamin** [comment]
sorry, did push to ceph-ci I think back when @tchaikov requested, but I've forced a rebased version there just now

**Comment by ljflores** [comment]
@mattbenjamin if this needs additional QA, please add the "needs-qa" label, and it will be picked up and run through the rados suite.

**Comment by mattbenjamin** [comment]
> @mattbenjamin if this needs additional QA, please add the "needs-qa" label, and it will be picked up and run through the rados suite.

sure, thank you @ljflores

**Comment by mattbenjamin** [comment]
(*I* don't think it needs-qa, but it should be harmless to include it for non-regression)

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/REEF#httpstrellocomcT9bt0Ezq1871-wip-yuri4-testing-2023-10-31-1447-old-wip-yuri4-testing-2023-10-30-1117-old-wip-yuri4-testing-2023-10-27-1200-old-wip-yuri4-test

**Comment by cbodley** [comment]
saw an odd failure in a recent arm build: https://jenkins.ceph.com/job/ceph-pull-requests-arm64/48235/consoleText
```
CMake Error at cmake/modules/BuildFIO.cmake:39 (set_target_properties):
  INTERFACE_LIBRARY targets may only have whitelisted properties.  The
  property "CXX_EXTENSIONS" is not allowed.
Call Stack (most recent call first):
  CMakeLists.txt:607 (build_fio)
```
that arm64 check failed here too but the log is gone

**Comment by cbodley** [comment]
not sure why arm64 is the only build failing there, unless it's running an older cmake version. i wasn't able to find a cmake version in the build log

**Pull request merged**